runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/stubs/ChannelCacheTest.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.dataflow.worker.windmill.client.grpc.stubs;
+
+import static org.junit.Assert.assertSame;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.function.Function;
+import org.apache.beam.runners.dataflow.worker.windmill.WindmillServiceAddress;
+import org.apache.beam.vendor.grpc.v1p60p1.io.grpc.ManagedChannel;
+import org.junit.After;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public class ChannelCacheTest {
+
+  private ChannelCache cache;
+
+  private static ChannelCache newCache(
+      Function<WindmillServiceAddress, ManagedChannel> channelFactory) {
+    return ChannelCache.forTesting(channelFactory, () -> {});
+  }
+
+  @After
+  public void cleanUp() {
+    if (cache != null) {
+      cache.clear();
+    }
+  }
+
+  private ManagedChannel newChannel(String channelName) {
+    return WindmillChannelFactory.inProcessChannel(channelName);
+  }
+
+  @Test
+  public void testLoadingCacheReturnsExistingChannel() {
+    String channelName = "existingChannel";
+    ManagedChannel channel = newChannel(channelName);
+    Function<WindmillServiceAddress, ManagedChannel> channelFactory =
+        spy(
+            new Function<WindmillServiceAddress, ManagedChannel>() {
+              @Override
+              public ManagedChannel apply(WindmillServiceAddress windmillServiceAddress) {
+                return channel;
+              }
+            });
+
+    cache = newCache(channelFactory);
+    WindmillServiceAddress someAddress = mock(WindmillServiceAddress.class);
+    // Initial call to load the cache.
+    cache.get(someAddress);
```

Comment:
might as well assert this is the same as channel

Reply:
done